dynamicbeat/pkg/checktypes/git/git.go

```diff
@@ -0,0 +1,146 @@
+package git
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"regexp"
+	"time"
+
+	"github.com/go-git/go-billy/v5/memfs"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/scorestack/scorestack/dynamicbeat/pkg/check"
+)
+
+// The Definition configures the behavior of the Git check and implements the "check" interface.
+type Definition struct {
+	Config          check.Config // Generic metadata about the check
+	Host            string       `optiontype:"required"`                    // IIP or FQDN the remote repository is located
+	Repository      string       `optiontype:"required"`                    // The path to the remote repository
+	Branch          string       `optiontype:"required"`                    // The branch to clone from the repository
+	Port            int          `optiontype:"optional"`                    // The port to connect to for cloning the repository
+	HTTPS           bool         `optiontype:"optional"`                    // Whether to use HTTP or HTTPS
+	HttpsValidate   bool         `optiontype:"optional"`                    // Whether HTTPS certificates should be validated
+	Username        string       `optiontype:"optional"`                    // Username to use for private repositories
+	Password        string       `optiontype:"optional"`                    // Password for the user
+	ContentMatch    bool         `optiontype:"optional"`                    // Whether to check the contents of a file
+	ContentFile     string       `optiontype:"optional"`                    // The path of the file to check the contents of
+	ContentRegex    string       `optiontype:"optional" optiondefault:".*"` // The regex to match against the checked file
+	CommitHashMatch bool         `optiontype:"optional"`                    // Whether or not to match the hash of the latest commit
+	CommitHash      string       `optiontype:"optional"`                    // The hash to check against the latest commit
+}
+
+// Run a single instance of the check.
+func (d *Definition) Run(ctx context.Context) check.Result {
+	// Initialilze empty result
+	result := check.Result{Timestamp: time.Now(), Metadata: d.Config.Metadata}
+
+	// Fail the check if an invalid method is attempted
+	var protocol string
+	switch d.HTTPS {
+	case true:
+		protocol = "https"
+	case false:
+		protocol = "http"
+	}
+
+	// Craft the repository url for git.Clone
+	// Using the implementation provided by go-git as it's probably better than anything I can come up with
+	repoUrl := transport.Endpoint{
+		Protocol: protocol,
+		Host:     d.Host,
+		Port:     d.Port,
+		User:     d.Username,
+		Password: d.Password,
+		Path:     d.Repository,
+	}
+
+	// Intialize memory storage for cloning
+	store := memory.NewStorage()
+	tree := memfs.New()
```

Comment:
Do these memory structures need to be closed or emptied in any way?

Reply:
The Filesystem interface does not provide a "close" method. The memory filesystem is considered garbage after the check thread is finished which is then cleaned up by the garbage collector.